hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/StakePeriodManager.java

```diff
@@ -0,0 +1,182 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.token.impl.handlers.staking;
+
+import static com.hedera.node.app.service.mono.utils.Units.MINUTES_TO_MILLISECONDS;
+import static com.swirlds.common.stream.LinkedObjectStreamUtilities.getPeriod;
+import static com.swirlds.common.utility.Units.MINUTES_TO_SECONDS;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.hedera.node.app.service.token.ReadableNetworkStakingRewardsStore;
+import com.hedera.node.config.ConfigProvider;
+import com.hedera.node.config.data.StakingConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * This class manages the current stake period and the previous stake period.
+ */
+@Singleton
+public class StakePeriodManager {
+    // Sentinel value for a field that wasn't applicable to this transaction
+    public static final long NA = Long.MIN_VALUE;
+    public static final ZoneId ZONE_UTC = ZoneId.of("UTC");
+    public static final long DEFAULT_STAKING_PERIOD_MINS = 1440L;
+
+    private final int numStoredPeriods;
+    private final long stakingPeriodMins;
+    private long currentStakePeriod;
+    private long prevConsensusSecs;
+
+    @Inject
+    public StakePeriodManager(@NonNull final ConfigProvider configProvider) {
+        final var config = configProvider.getConfiguration().getConfigData(StakingConfig.class);
+        numStoredPeriods = config.rewardHistoryNumStoredPeriods();
+        stakingPeriodMins = config.periodMins();
+    }
+
+    /**
+     * Returns the epoch second at the start of the given stake period. It is used in
+     * {@link com.hedera.node.app.service.token.impl.utils.RewardCalculator} to set the stakePeriodStart
+     * on each {@link com.hedera.hapi.node.state.token.StakingNodeInfo} object
+     * @param stakePeriod the stake period
+     * @return the epoch second at the start of the given stake period
+     */
+    public long epochSecondAtStartOfPeriod(final long stakePeriod) {
+        if (stakingPeriodMins == DEFAULT_STAKING_PERIOD_MINS) {
+            return LocalDate.ofEpochDay(stakePeriod).atStartOfDay().toEpochSecond(ZoneOffset.UTC);
+        } else {
+            return stakePeriod * stakingPeriodMins * MINUTES_TO_SECONDS;
+        }
+    }
```

Comment:
It'd be nice to start using the java time api instead of longs and figuring out times ourselves, since things like daylight savings can often mess it up.

That said, this is a copy-paste, so we can address later